### PR TITLE
Bar plot performance

### LINKED ISSUE #1289: Gridlines on Scale.Category

### PR DESCRIPTION
**Pull request by aicioara**
Closes https://github.com/palantir/plottable/issues/1284

Improved the call of `flush()` from `132 ms` down to `10 ms` by caching the result of `_getBarPixelWidth()`

Historic Before: http://jsfiddle.net/aycxz2ap/4/
Before: http://jsfiddle.net/aycxz2ap/1/
After: http://jsfiddle.net/aycxz2ap/3/

**QE Notes:** Bar Plot should behave the same. In particular the width of the bars should not behave weirdly. Also The main performance boost is in the PanZoom animation effect (as can be seen in the fiddle)

### CODE REVIEW

File: src/plots/barPlot.ts [hunk 1]
```diff
@@ -27,6 +27,9 @@ export module Plots {
     private _labelConfig: Utils.Map<Dataset, LabelConfig>;
     private _baselineValueProvider: () => (X|Y)[];
 
+    private _barPixelWidth = 0;
+    private _updateBarPixelWidthCallback: (() => void);
```
Comment (bluong) [inline src/plots/barPlot.ts]: I think? the parenthese for `() => void` is unnecessary?

File: src/plots/barPlot.ts [hunk 2]
```diff
@@ -479,6 +505,7 @@ export module Plots {
      * If the position scale of the plot is a QuantitativeScale, then _getMinimumDataWidth is scaled to compute the barPixelWidth
      */
     protected _getBarPixelWidth(): number {
+
```
Comment (bluong) [inline src/plots/barPlot.ts]: I don't think the whitespace here is necessary

File: src/plots/barPlot.ts [hunk 3]
```diff
@@ -495,14 +522,14 @@ export module Plots {
 
         numberBarAccessorData.sort((a, b) => a - b);
 
-        var barAccessorDataPairs = d3.pairs(numberBarAccessorData);
+        var scaledData = numberBarAccessorData.map((datum: number) => barScale.scale(datum));
```
Comment (bluong) [inline src/plots/barPlot.ts]: Can you remove the `number` type here or no?

File: src/plots/barPlot.ts [hunk 4]
```diff
@@ -518,6 +545,10 @@ export module Plots {
       return barPixelWidth;
     }
 
+    private _updateBarPixelWidth() {
+      this._barPixelWidth = this._getBarPixelWidth();
```
Comment (bluong) [inline src/plots/barPlot.ts]: Why do we need the function wrapping around this line?

File: src/plots/barPlot.ts [hunk 5]
```diff
@@ -553,6 +584,11 @@ export module Plots {
       return { x: x, y: y };
     }
 
+  protected _uninstallScaleForKey(scale: Scale<any, number>, key: string) {
+    scale.offUpdate(this._updateBarPixelWidthCallback);
+    super._uninstallScaleForKey(scale, key);
+  }
```
Comment (bluong) [inline src/plots/barPlot.ts]: Indentation is messed up

File: src/plots/barPlot.ts (hunk 4 above)
Comment (aicioara) [inline src/plots/barPlot.ts]: Backward compatibility

Comment (aicioara) [inline src/plots/barPlot.ts]: if I add that line in the `_getBarPixelWidth()` it does not really make much sense from a reading stand point. Also, can't rename or remove old function

**Pull request merged**